From c0ecdeafe3c69487ad3749c6c1c473ab3d88f521 Mon Sep 17 00:00:00 2001
From: mohit pal singh <mohitpal@qti.qualcomm.com>
Date: Tue, 6 Feb 2024 13:01:10 -0800
Subject: [PATCH] squashfs-tools: Allow setting selinux xattrs through
 file_context

Add a context-file flag that allows passing an selinux
security context file to set security.selinux xattrs
rather than reading xattrs from filesystem's
source directory.

Port the change to LE environment. Change the path provided
to selabel_lookup() to use the path from the root of the
filesystem. Add WITH_SELINUX to CFLAGS to ensure the code
compiles fine even when selinux
library support is not available.
---
 squashfs-tools/Makefile     | 16 +++++--
 squashfs-tools/mksquashfs.c | 11 ++++-
 squashfs-tools/xattr.c      | 86 ++++++++++++++++++++++++++++++++++++-
 3 files changed, 106 insertions(+), 7 deletions(-)
 mode change 100755 => 100644 squashfs-tools/Makefile

diff --git a/squashfs-tools/Makefile b/squashfs-tools/Makefile
old mode 100755
new mode 100644
index 5795d0d0..7c2a8eda
--- a/squashfs-tools/Makefile
+++ b/squashfs-tools/Makefile
@@ -169,6 +169,12 @@ CFLAGS += $(EXTRA_CFLAGS) $(INCLUDEDIR) -D_FILE_OFFSET_BITS=64 \
 	-Wall
 
 LIBS = -lpthread -lm
+
+ifeq ($(WITH_SELINUX),1)
+CFLAGS += -DWITH_SELINUX
+LIBS += -lselinux
+endif
+
 ifeq ($(GZIP_SUPPORT),1)
 CFLAGS += -DGZIP_SUPPORT
 MKSQUASHFS_OBJS += gzip_wrapper.o
diff --git a/squashfs-tools/mksquashfs.c b/squashfs-tools/mksquashfs.c
index aaa4b002..dfeb2214 100644
--- a/squashfs-tools/mksquashfs.c
+++ b/squashfs-tools/mksquashfs.c
@@ -130,6 +130,7 @@ unsigned int cache_bytes = 0, cache_size = 0, inode_count = 0;
 /* inode lookup table */
 squashfs_inode *inode_lookup_table = NULL;
 
+char *context_file = NULL;
 struct inode_info *inode_info[INODE_HASH_SIZE];
 
 /* hash tables used to do fast duplicate searches in duplicate check */
@@ -6094,6 +6095,7 @@ static void print_sqfstar_options(FILE *stream, char *name, int total_mem)
 	fprintf(stream, "-no-sparse\t\tdon't detect sparse files\n");
 	fprintf(stream, "-no-xattrs\t\tdon't store extended attributes" NOXOPT_STR "\n");
 	fprintf(stream, "-xattrs\t\t\tstore extended attributes" XOPT_STR "\n");
+	fprintf(stream, "-context-file <file>\tApply selinux security xattrs from context-file instead of reading xattrs from file system\n");
 	fprintf(stream, "-noI\t\t\tdo not compress inode table\n");
 	fprintf(stream, "-noId\t\t\tdo not compress the uid/gid table (implied by ");
 	fprintf(stream, "-noI)\n");
@@ -6651,7 +6653,14 @@ print_sqfstar_compressor_options:
 				exit(1);
 			}
 
-		} else if(strcmp(argv[i], "-nopad") == 0)
+		} else if(strcmp(argv[i], "-context-file") == 0) {
+			if(++i == argc) {
+				ERROR("%s: -context-file: missing file name\n",
+					argv[0]);
+				exit(1);
+			}
+		}
+		else if(strcmp(argv[i], "-nopad") == 0)
 			nopad = TRUE;
 
 		else if(strcmp(argv[i], "-info") == 0)
diff --git a/squashfs-tools/xattr.c b/squashfs-tools/xattr.c
index 8cfb8d50..82b08c95 100644
--- a/squashfs-tools/xattr.c
+++ b/squashfs-tools/xattr.c
@@ -36,6 +36,11 @@
 #include <stdlib.h>
 #include <sys/xattr.h>
 
+#ifdef WITH_SELINUX
+#include <selinux/label.h>
+#include <selinux/selinux.h>
+#endif
+
 #include "squashfs_fs.h"
 #include "squashfs_swap.h"
 #include "mksquashfs.h"
@@ -45,6 +50,8 @@
 #include "pseudo.h"
 #include "tar.h"
 
+#define ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
+
 /* compressed xattr table */
 static char *xattr_table = NULL;
 static unsigned int xattr_size = 0;
@@ -78,6 +85,7 @@ extern int no_xattrs, noX;
 extern long long bytes;
 extern int fd;
 extern unsigned int xattr_bytes, total_xattr_bytes;
+extern char *context_file;
 
 /* helper functions from mksquashfs.c */
 extern unsigned short get_checksum(char *, int, unsigned short);
@@ -112,7 +120,69 @@ int xattr_get_prefix(struct xattr_list *xattr, char *name)
 	return prefix_table[i].type;
 }
 
+#ifdef WITH_SELINUX
+
+static struct selabel_handle *sehnd = NULL;
+
+struct selabel_handle *get_sehnd(const char *context_file) {
+    struct selinux_opt seopts[] = {
+        {
+            .type = SELABEL_OPT_PATH,
+            .value = context_file
+        }
+    };
+    struct selabel_handle *sehnd =
+        selabel_open(SELABEL_CTX_FILE, seopts, ARRAY_SIZE(seopts));
+
+    if (!sehnd) {
+        perror("Error running selabel_open.");
+        exit(EXIT_FAILURE);
+    }
+    return sehnd;
+}
+
+char *set_selabel(const char *path, unsigned int mode, struct selabel_handle *sehnd) {
+    char *secontext;
+    if (sehnd != NULL) {
+        int full_name_size = strlen(path) + 2;
+        char* full_name = (char*) malloc(full_name_size);
+        if (full_name == NULL) {
+            perror("Malloc Failure.");
+            exit(EXIT_FAILURE);
+        }
+
+        full_name[0] = '/';
+        strncpy(full_name + 1, path, full_name_size - 1);
+
+        if (selabel_lookup(sehnd, &secontext, full_name, mode)) {
+            secontext = strdup("u:object_r:unlabeled:s0");
+        }
+
+        free(full_name);
+        return secontext;
+    }
+    perror("Selabel handle is NULL.");
+    exit(EXIT_FAILURE);
+}
+
 	
+static int read_xattrs_from_context_file(char *filename, int mode,
+	struct selabel_handle *sehnd, struct xattr_list **xattrs)
+{
+	char *attr_val;
+	struct xattr_list *x = malloc(sizeof(*x));
+	if(x == NULL)
+		MEM_ERROR();
+
+	x->type = get_prefix(x, "security.selinux");
+	attr_val = set_selabel(filename, mode, sehnd);
+	x->value = (void *)attr_val;
+	x->vsize = strlen(attr_val);
+	*xattrs = x;
+	return 1;
+}
+#endif
+
 static int read_xattrs_from_system(char *filename, struct xattr_list **xattrs)
 {
 	ssize_t size, vsize;
@@ -611,11 +681,13 @@ int generate_xattrs(int xattrs, struct xattr_list *xattr_list)
 }
 
 
+extern char *subpathname(struct dir_ent *dir_ent);
 int read_xattrs(void *d)
 {
 	struct dir_ent *dir_ent = d;
 	struct inode_info *inode = dir_ent->inode;
 	char *filename = pathname(dir_ent);
+	char *subpath = subpathname(dir_ent);
 	struct xattr_list *xattr_list;
 	int xattrs;
 
@@ -624,8 +696,18 @@ int read_xattrs(void *d)
 
 	if(IS_TARFILE(inode))
 		xattrs = read_xattrs_from_tarfile(inode, &xattr_list);
-	else
-		xattrs = read_xattrs_from_system(filename, &xattr_list);
+	else {
+			#ifdef WITH_SELINUX
+				if (context_file) {
+					if (sehnd == NULL)
+						sehnd = get_sehnd(context_file);
+						xattrs = read_xattrs_from_context_file(subpath, inode->buf.st_mode,
+						sehnd, &xattr_list);
+					} else {
+			#else
+				xattrs = read_xattrs_from_system(filename, &xattr_list);
+			#endif
+	}
 	if(xattrs == 0)
 		return SQUASHFS_INVALID_XATTR;
 
-- 
2.17.1

